### PR TITLE
feat: update csv reading/writing infrastructure to new digitization

### PR DESCRIPTION
**Pull request by asalzburger**
This PR replaces #694 and #700.

It follows the removal of the `FatrasDigitization` and introduces a new `ActsExamples/Digitization` which should in a follow-up PR be re-used in the reconstruction examples.

It also adds a round-trip test for writing / reading measurements.

### CODE REVIEW

**Review by XiaocongAi — COMMENTED** [review]
Hi @asalzburger , it's great to see a standalone digitization example. The PR looks clear to me, but I still have a few questions about the I/O logic.

File: Examples/Run/Digitization/Common/DigitizationExample.cpp [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+  // Read the sim hits
+  auto simHitReaderCfg = setupSimHitReading(vm, sequencer);
+  // Read the particles
+  auto particleReaderCfg = setupParticleReading(vm, sequencer);
+
+  auto cfile = vm["digi-config-file"].as<std::string>();
+
+  if (not cfile.empty() or vm["digi-smear"].as<bool>()) {
+    // To be handles differently between DigitizationAlgorithm/SmearingAlgorithm
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/Common/DigitizationExample.cpp:96]: Thanks.

File: Examples/Run/Digitization/Common/DigitizationExample.cpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  }
+
+  // Setup the magnetic field
+  Options::setupMagneticFieldServices(vm, sequencer);
+  auto magneticField = Options::readMagneticField(vm);
+
+  // Read the sim hits
+  auto simHitReaderCfg = setupSimHitReading(vm, sequencer);
+  // Read the particles
+  auto particleReaderCfg = setupParticleReading(vm, sequencer);
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/Common/DigitizationExample.cpp:91]: Yes, I guess we need to link them together, that's I wanted to check with @msmk0 though.

File: Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    /// Which cluster collection to write (optional)
+    std::string inputClusters;
+    /// Which simulated (truth) hits collection to use.
+    std::string inputSimHits;
+    /// Input collection to map measured hits to simulated hits.
+    std::string inputMeasurementSimHitsMap;
+    /// Where to place output files
+    std::string outputDir;
+    /// Write the optional measurement surfaces collection
+    bool writeMeasurementSurfaces = false;
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp:64]: Good call, I will add that, in fact.

File: Examples/Run/Digitization/Common/DigitizationExample.cpp [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        if (vm["digi-read-write-test"].as<bool>()) {
+          // Read truth hits from CSV files
+          auto measReaderCfg = Options::readCsvMeasurementReaderConfig(vm);
+          measReaderCfg.inputDir = vm["input-dir"].as<std::string>();
+          measReaderCfg.outputMeasurements = outputMeasurements;
+          measReaderCfg.outputClusters = outputClusters;
+          measReaderCfg.outputMeasurementIds = "measurementids";
+          measReaderCfg.outputMeasurementParticlesMap =
+              outputMeasurementParticlesMap;
+          measReaderCfg.outputSimHits = simHitReaderCfg.outputSimHits;
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/Common/DigitizationExample.cpp:130]: Added a check.

File: Examples/Io/Csv/src/CsvMeasurementReader.cpp [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    // Map internal measurement index back to original, non-monotonic hit id
+    measurementIds.push_back(m.hit_id);
+  }
+
+  // Write the data to the EventStore
+  ctx.eventStore.add(m_cfg.outputMeasurements, std::move(measurements));
+  ctx.eventStore.add(m_cfg.outputMeasurementIds, std::move(measurementIds));
+  ctx.eventStore.add(m_cfg.outputClusters, std::move(clusters));
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Io/Csv/src/CsvMeasurementReader.cpp:307]: Added a check here.

File: Examples/Io/Csv/src/CsvMeasurementReader.cpp [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+  std::vector<ActsExamples::CellData> cellData = {};
+  if (not m_cfg.outputClusters.empty()) {
+    cellData = readCellsByHitId(m_cfg.inputDir, ctx.eventNumber);
+  }
+  auto truthData = readTruthHitsByHitId(m_cfg.inputDir, ctx.eventNumber);
+
+  // Prepare containers for the hit data using the framework event data types
+  GeometryIdMultimap<Measurement> measurements;
+  ClusterContainer clusters;
+  IndexSourceLinkContainer sourceLinks;
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Io/Csv/src/CsvMeasurementReader.cpp:178]: Again, I think the reader is not yet fully complete - I wanted to get some insight from @msmk0

File: Examples/Run/Digitization/Common/DigitizationInput.hpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#include "ActsExamples/Io/Csv/CsvParticleReader.hpp"
+#include "ActsExamples/Io/Csv/CsvSimHitReader.hpp"
+#include "ActsExamples/Utilities/Options.hpp"
+
+#include <memory>
+#include <string>
+
+/// Setup sim hit csv reader
+///
+/// @param variables The configuration variables
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/Common/DigitizationInput.hpp:24]: Fixed it in several other occasions

File: Examples/Run/Digitization/Common/DigitizationExample.cpp [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        in >> djson;
+        in.close();
+
+        outputClusters = "clusters";
+
+        DigiConfigContainer digitizationConfigs =
+            DigiConfigConverter("digitization-configuration").fromJson(djson);
+
+        if (vm["digi-read-write-test"].as<bool>()) {
+          // Read truth hits from CSV files
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/Common/DigitizationExample.cpp:122]: Fixed.

File: Examples/Run/Digitization/Common/DigitizationExample.hpp [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+namespace ActsExamples {
+class IBaseDetector;
+}
+
+/// The Propagation example
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/Common/DigitizationExample.hpp:15]: Fixed.

File: Examples/Run/Digitization/GenericDigitizationExample.cpp [hunk 10]
```diff
@@ -0,0 +1,16 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/GenericDigitizationExample.cpp:3]: Fixed.

File: Examples/Run/Digitization/Common/DigitizationExample.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/Common/DigitizationExample.cpp:130]: For the moment, that's the only way I can get the right name, as discussed, there will be a follow-up PR with container name handling.

File: Examples/Run/Digitization/Common/DigitizationInput.cpp [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+ActsExamples::CsvParticleReader::Config setupParticleReading(
+    const ActsExamples::Options::Variables& vars,
+    ActsExamples::Sequencer& sequencer) {
+  using namespace ActsExamples;
+
+  // Read some standard options
+  auto logLevel = Options::readLogLevel(vars);
+
+  // Read particles (initial states) and clusters from CSV files
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/Common/DigitizationInput.cpp:42]: Fixed.

File: Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp]: Actually I scraped this, we don't need that.

File: Examples/Run/Digitization/Common/DigitizationInput.hpp [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+/// @param sequencer The framework sequencer
+///
+/// @return config for sim hits csv reader
+ActsExamples::CsvSimHitReader::Config setupSimHitReading(
+    const ActsExamples::Options::Variables& vars,
+    ActsExamples::Sequencer& sequencer);
+
+/// Setup sim particle csv reader
+///
+/// @param variables The configuration variables
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/Common/DigitizationInput.hpp]: Fixed.

File: Examples/Run/Digitization/Common/DigitizationExample.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Examples/Run/Digitization/Common/DigitizationExample.cpp:91]: For the linking, the particle identifier should be sufficient. The simulation hit also stores the particle state at the detection, so the initial kinematic information is not required either.

File: Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementReader.hpp [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    /// Where to read input files from.
+    std::string inputDir;
+    /// Output measurement collection.
+    std::string outputMeasurements;
+    /// Output cluster collection (optional).
+    std::string outputClusters;
+    /// For each cluster/ hit index the original hit id stored on file.
+    std::string outputMeasurementIds;
+    /// Output hit-particles mapping collection.
+    std::string outputMeasurementParticlesMap;
```
Comment (msmk0) [inline Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementReader.hpp:53]: Naively, I would have expected to get a measurement/simulation hit map **not** a measurement/particle map. That way, only the data related directly for this steps needs to be stored, i.e.
```
Simulation -> SimHit/Particle map
Digitiation -> Measurement/SimHit map
```
If I need the Measurement/Particle map, I can always generate it from the two separate maps (at which point both the measurements, simulation hits, and particles need to be read).

File: Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementReader.hpp [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    /// Output simulated (truth) hits collection.
+    std::string outputSimHits;
+    /// Tracking geometry required to access global-to-local transforms.
+    std::shared_ptr<const Acts::TrackingGeometry> trackingGeometry;
```
Comment (msmk0) [inline Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementReader.hpp:57]: Same question as for the writer: the measurement information is purely local, why are the transformation required?

File: Examples/Io/Csv/src/CsvMeasurementReader.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Examples/Io/Csv/src/CsvMeasurementReader.cpp:178]: I do not think it is needed.

File: Examples/Io/Csv/src/CsvOutputData.hpp [hunk 15]
```diff
@@ -114,21 +114,39 @@ struct HitData {
                  x, y, z, t);
 };
 
+struct MeasurementData {
+  /// Event-unique measurement identifier. Each value can appear at most once.
+  uint64_t hit_id;
+  /// Hit surface identifier.
+  uint64_t geometry_id = 0u;
+  /// Partially decoded hit surface identifier components.
+  uint32_t volume_id, layer_id, module_id;
```
Comment (msmk0) [inline Examples/Io/Csv/src/CsvOutputData.hpp:123]: Do we really want to store this? It was in the TrackML dataset only to help the participants. It is duplicated information for the purpose of data storage.

File: Examples/Io/Csv/src/CsvOutputData.hpp [hunk 16]
```diff
@@ -114,21 +114,39 @@ struct HitData {
...
+  /// Local hit information - bit identification what's measured
+  uint8_t local_key;
+  float local0, local1, phi, theta, time;
+  float varLocal0, varLocal1, varPhi, varTheta, varTime;
```
Comment (msmk0) [inline Examples/Io/Csv/src/CsvOutputData.hpp:127]:
```suggestion
  float var_local0, var_local1, var_phi, var_theta, var_time;
```
Consistency with the rest of the CSV files.

File: Examples/Io/Csv/src/CsvOutputData.hpp [hunk 17]
```diff
@@ -114,21 +114,39 @@ struct HitData {
                  x, y, z, t);
 };
 
+struct MeasurementData {
+  /// Event-unique measurement identifier. Each value can appear at most once.
+  uint64_t hit_id;
```
Comment (msmk0) [inline Examples/Io/Csv/src/CsvOutputData.hpp:119]: This should probably be `measurement_id`.

**Review by msmk0 — COMMENTED** [review]
@asalzburger I have the feeling there is quite a bit of code duplication here, e.g. the measurement readers are also writing out simulated hits again.
- There is no need for the `-truth.csv` file anymore. This information has been superseded by the SimHits data that is stored by the fatras example as `-hits.csv` (or `-simhits.csv`, I am not quite sure). All the information is already written out by the simulation and there I see no need to write it out again for the digitization.
- Within TrackML we had an explicit `hit_id` column to connect truth to spacepoints. This only allows 1-1 association and we documented it as being an arbitrary number. In reality, it was the hit index in the container with some special value (0) to mark noise hits without associated truth. `hit_id` should go away completely.
- Instead, we need to support n-m association, i.e. arbitrary association between sim hits and measurements. The best way to handle this is to write out a map file that stores tuples of sim hit and measurement identifiers.
- There are two options for the identifiers: do not write out any and use the position in the container/file as the identifier. Then the map file should have columns `measurement_index,simhit_index` or `..._idx` if you prefer shorter ones. Or you write out an explicit identifier column for both simulated hits and measurements; columns names should then be `simhit_id` and `measurement_id` and they should be treated as opaque values similar to `particle_id` for consistency. I think, the index variant is easier to implement and maps closer to how we handle containers in the examples.

As already mentioned by @XiaocongAi the measurement reader/writer should not care about the simulated hits, the global/local transformations, or the particles.

**Comment by asalzburger** [comment]
Thanks for all your comments @msmk0. I will incorporate them, the less writing, the better!

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/MeasurementCreation.hpp [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Definitions/TrackParametrization.hpp"
+#include "ActsExamples/EventData/Cluster.hpp"
+#include "ActsExamples/EventData/IndexSourceLink.hpp"
+#include "ActsExamples/EventData/Measurement.hpp"
+
+namespace ActsExamples {
+
+/// Nested struct for digitized parameters
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/MeasurementCreation.hpp:19]: Fixed.

File: Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+#include "ActsExamples/Framework/WriterT.hpp"
+
+#include <memory>
+#include <mutex>
+#include <vector>
+
+#include <TTree.h>
+
+class TFile;
+class TTree;
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp:33]: Fixed.

File: Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+#include "ActsExamples/Digitization/SmearingAlgorithm.hpp"
+#include "ActsExamples/EventData/Cluster.hpp"
+#include "ActsExamples/EventData/Index.hpp"
+#include "ActsExamples/EventData/Measurement.hpp"
+#include "ActsExamples/EventData/SimHit.hpp"
+#include "ActsExamples/Framework/WriterT.hpp"
+
+#include <memory>
+#include <mutex>
+#include <vector>
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp:28]: Fixed.

File: Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+ public:
+  struct Config {
+    /// Which measurement collection to write.
+    std::string inputMeasurements;
+    /// Which cluster collection to write (optional)
+    std::string inputClusters;
+    /// Which simulated (truth) hits collection to use.
+    std::string inputSimHits;
+    /// Input collection to map measured hits to simulated hits.
+    std::string inputMeasurementSimHitsMap;
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp:61]: Removed.

File: Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+    /// Where to place output files
+    std::string outputDir;
+    /// Number of decimal digits for floating point precision in output.
+    size_t outputPrecision = std::numeric_limits<float>::max_digits10;
+    /// Tracking geometry required to access global-to-local transforms.
+    std::shared_ptr<const Acts::TrackingGeometry> trackingGeometry = nullptr;
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementWriter.hpp:67]: Correct, removed.

File: Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementReader.hpp [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+ public:
+  struct Config {
+    /// Where to read input files from.
+    std::string inputDir;
+    /// Output measurement collection.
+    std::string outputMeasurements;
+    /// Output cluster collection (optional).
+    std::string outputClusters;
+    /// For each cluster/ hit index the original hit id stored on file.
+    std::string outputMeasurementIds;
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementReader.hpp:51]: Removed.

File: Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementReader.hpp (hunk 14 above)
Comment (asalzburger) [reply Examples/Io/Csv/include/ActsExamples/Io/Csv/CsvMeasurementReader.hpp:57]: Removed now.

File: Examples/Io/Csv/src/CsvOutputData.hpp (hunk 15 above)
Comment (asalzburger) [reply Examples/Io/Csv/src/CsvOutputData.hpp:123]: Happy to go back to geometry_id only.

File: Examples/Io/Csv/src/CsvOutputData.hpp (hunk 16 above)
Comment (asalzburger) [reply Examples/Io/Csv/src/CsvOutputData.hpp]: Yup, done.

File: Examples/Io/Csv/src/CsvOutputData.hpp (hunk 17 above)
Comment (asalzburger) [reply Examples/Io/Csv/src/CsvOutputData.hpp:116]: Done.

File: Examples/Run/Digitization/Common/DigitizationExample.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Run/Digitization/Common/DigitizationExample.cpp:91]: I changed to write out the measurement-simhit map.

File: Examples/Io/Csv/src/CsvPlanarClusterReader.cpp [hunk 24]
```diff
@@ -79,16 +79,7 @@ struct CompareHitId {
 /// Convert separate volume/layer/module id into a single geometry identifier.
 inline Acts::GeometryIdentifier extractGeometryId(
     const ActsExamples::HitData& data) {
-  // if available, use the encoded geometry directly
```
Comment (asalzburger) [inline Examples/Io/Csv/src/CsvPlanarClusterReader.cpp:82]: As the PlanarClusterReader will vanish, this fulfills the right functionality now, but I didn't really do it cleanly as this code will go dead soon ...

**Comment by asalzburger** [comment]
Hi @XiaocongAi - would you mind having a second look over it?

File: Examples/Io/Csv/src/CsvMeasurementReader.cpp [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+      readMeasurementsByGeometryId(m_cfg.inputDir, ctx.eventNumber);
+
+  std::vector<ActsExamples::CellData> cellData = {};
+  if (not m_cfg.outputClusters.empty()) {
+    cellData = readCellsByHitId(m_cfg.inputDir, ctx.eventNumber);
+  }
+
+  // Prepare containers for the hit data using the framework event data types
+  GeometryIdMultimap<Measurement> measurements;
+  ClusterContainer clusters;
```
Comment (XiaocongAi) [inline Examples/Io/Csv/src/CsvMeasurementReader.cpp:133]: Is the `clusters` filled anywhere?

File: Examples/Run/Digitization/Common/DigitizationExample.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (XiaocongAi) [reply Examples/Run/Digitization/Common/DigitizationExample.cpp:91]: Except the particle id, we are not really using any particle info, right? To me, this truth particles reading could be removed.

**Review by XiaocongAi — COMMENTED** [review]
Hi @asalzburger , thank you for the update. It looks nice to me, but there are two things I'd like to confirm with you in case I missed something:

- I would expect the `CsvMeasurementReader` will create the `clusters` if the `clusters` collection name is not empty. But I don't see the code there to create the clusters. Is it planned to be addressed by a follow-up PR?
- In the `DigitizationExample`, I think we don't have to read in the particles since the sim particles are not really used anywhere. But maybe I'm wrong?

We might discuss these tomorrow.

**Review by XiaocongAi — APPROVED** [review]

**Comment by XiaocongAi** [comment]
OK, let's get this in and address them in a following-up PR.

**Pull request merged**